x-pack/legacy/plugins/ml/public/data_frame/pages/data_frame_new_pivot/components/step_define/step_define_form.tsx

```diff
@@ -450,31 +529,131 @@ export const StepDefineForm: SFC<Props> = React.memo(({ overrides = {}, onChange
+                      helpText={i18n.translate('xpack.ml.dataframe.stepDefineForm.queryHelpText', {
+                        defaultMessage: 'Use a query to filter the source data (optional).',
+                      })}
+                    >
+                      <KqlFilterBar
+                        indexPattern={indexPattern}
+                        onSubmit={searchHandler}
+                        initialValue={searchString === defaultSearch ? emptySearch : searchString}
+                        placeholder={i18n.translate(
+                          'xpack.ml.dataframe.stepDefineForm.queryPlaceholder',
+                          {
+                            defaultMessage: 'e.g. {example}',
+                            values: { example: 'method : "GET" or status : "404"' },
+                          }
+                        )}
+                      />
+                    </EuiFormRow>
+                  )}
+                </Fragment>
+              )}
+            </Fragment>
+          )}
+
+          {isAdvancedSourceEditorEnabled && (
+            <Fragment>
+              <EuiFormRow
+                label={i18n.translate(
+                  'xpack.ml.dataframe.stepDefineForm.advancedSourceEditorLabel',
+                  {
+                    defaultMessage: 'Source configuration object',
+                  }
+                )}
+                helpText={advancedSourceEditorHelpText}
+              >
+                <EuiPanel grow={false} paddingSize="none">
+                  <EuiCodeEditor
+                    mode="json"
+                    width="100%"
+                    value={advancedEditorSourceConfig}
+                    onChange={(d: string) => {
+                      setAdvancedEditorSourceConfig(d);
+
+                      // Disable the "Apply"-Button if the config hasn't changed.
+                      if (advancedEditorSourceConfigLastApplied === d) {
+                        setAdvancedSourceEditorApplyButtonEnabled(false);
+                        return;
+                      }
+
+                      // Try to parse the string passed on from the editor.
+                      // If parsing fails, the "Apply"-Button will be disabled
+                      try {
+                        JSON.parse(d);
+                        setAdvancedSourceEditorApplyButtonEnabled(true);
+                      } catch (e) {
+                        setAdvancedSourceEditorApplyButtonEnabled(false);
+                      }
+                    }}
+                    setOptions={{
+                      fontSize: '12px',
+                    }}
+                    aria-label={i18n.translate(
+                      'xpack.ml.dataframe.stepDefineForm.advancedSourceEditorAriaLabel',
                       {
-                        defaultMessage: 'e.g. {example}',
-                        values: { example: 'method : "GET" or status : "404"' },
+                        defaultMessage: 'Advanced editor',
```

Comment:
As @walterra suggested, probably better to give the two advanced editors separate labels. Advanced query editor', 'Advanced pivot editor' perhaps?

Reply:
Good idea - updated in 261ed3bc5b2a487226934e5f31aa2f5c9c3ebf1f